lib/std/Progress.zig

```diff
@@ -143,7 +143,7 @@ pub const Node = struct {
 /// API to return Progress rather than accept it as a parameter.
 /// `estimated_total_items` value of 0 means unknown.
 pub fn start(self: *Progress, name: []const u8, estimated_total_items: usize) *Node {
-    const stderr = std.io.getStdErr();
+    var stderr: std.io.StdErrType = std.io.getStdErr();
```

Comment:
Similar thing here, if these have to be `var` now instead of `const` then I imagine it would have to be done in other places too, all just because of Windows. I hope we can find a solution where there are no external changes needed.

Reply:
Maybe this is just me getting mixed up between `const` and `var` keywords in Zig world, but wouldn't it be bad form/illegal to change the state of a field within a `const` like with the rotating fifo_utf8 buffer needed for reads? If it is, wouldn't that mean these have to be a `var` anyways if I introduce a change to `File` to read Utf16Le on Windows consoles?

I can try experimenting with changing `File` read/write to dance around Windows console. It would get rid of the need to update a lot of the compiler code to make it work within this repo.

Now this is if Zig does the translating of Utf16Le -> Utf8. If we have Windows do it under the hood, it would get rid of the need to change `File` or how we get Std handles. I'm just currently avoiding it to prevent any potential harm of changing someone's code page without them knowing.

Reply:
Hmm I just really don't think that whatever `std.io.getStdErr` returns should have state like that such that you need to make it `var` in order to write or read from it. `std.fs.File` is a representation of the file, not the file itself.
So I think this change is just kind of breaking that way of thinking about it. I think it should keep returning always only one struct (`std.fs.File`).

But even so, anything that can be `const` should be `const`. In the future using `var` when you could use `const` might become an error: https://github.com/ziglang/zig/issues/224 which is supposed to remind the developer that maybe they wanted to mutate that variable.

> I can try experimenting with changing `File` read/write to dance around Windows console. It would get rid of the need to update a lot of the compiler code to make it work within this repo.

Yes, I think that's the right approach.

> If we have Windows do it under the hood, it would get rid of the need to change `File` or how we get Std handles.

That's possible with the `GetConsoleMode` and `SetConsoleOutputCP` functions you mentioned right? Hmm well for buffered writing for example you could set the output code page to UTF-8 and then after the flush reset it.
Well, I think there will be many people that will always want to use UTF-8 throughout their whole application, so maybe we just set the output code page to UTF-8 at the start of the program (or at the first write of anything) using `std.once` or something and then if the user wants any other code page they have to set it themselves? We could expose `SetConsoleOutputCP` with a nicer API or something.

However if the user sets the code page and only after that does their first write, the code page will be set to UTF-8 again. Maybe it is better to do it on program startup. But what if they never write anything in their whole program?
I don't know if you have any ideas here?

So maybe it is better to follow suit and do it without using those functions, like Rust and Go.

Reply:
Been messing around with this off and on over the weekend. I can definitely get `write` to be solely in `File`. `read` on the other hand is being a bit tricky here, as I need mutability somewhere within itself or how it returns a reader. Unfortunately, most everything that uses `File` will use it with `const` for both it and the reader.

I've tried being a bit sneaky with how to return the reader:
```zig
    pub const WinConsoleReader = io.ConsoleReader(4096);
    pub const Reader = if (is_windows) WinConsoleReader.Reader else io.Reader(File, ReadError, read);

    pub fn reader(file: File) Reader {
        if (is_windows) {
            return (WinConsoleReader{ .file = file }).reader();
```

However, this is probably a footgun of some sorts where the scope of the WinConsoleReader is ended, so the reader ends up pointing to junk. Got any ideas as to how I might be able to utilize a small mutable buffer for reads without changing the current APIs?

Reply:
Hmm if I look at Rust, it does seem to work like `let mut file = File::create("foo.txt")?;` there (so `var` instead of `let mut` in Zig) so maybe we will have to change all `const`s to `var`s or something? In that case it might make sense to discuss this further in a separate issue.

I'm still wondering if this is a feasible approach?

> Perhaps we can disallow (make it "unsafe") to set the code page with `SetConsoleOutputCP` and they have to use our function so that we can keep track of whether the user already set an explicit code page or not. If not, we can set it to UTF-8 on the first write of anything. Does this sound possible?

It does sound pretty simple to me. But maybe it's too hacky to do.

I suppose `GetConsoleOutputCP` or something would not be of much use because I assume these functions return UTF-16 as the default before any code page has been set.

Reply:
> And ultimately, using var instead of const would only be required on Windows anyway, I think. I'm just thinking it might be a bit inconsistent when you're reading code but that's not very important.

If that feature comes into play where it complains about users using `var` instead of `const`, it will cause compilation errors for non-Windows targets, if they are developing from a Windows machine.

> It does sound pretty simple to me. Also sounds kinda hacky though so it's probably not the best solution.
>
> I suppose GetConsoleOutputCP or something would not be of much use because I assume these functions return UTF-16 as the default before any code page has been set (so that we can check whether an explicit code page has been set).

While I think we could find a way to use those API sets, I fear it may cause the active session's code page to change and there be no way back to the original code page. Sure we could try and gather what the current code page is and set it back at the end, but if the program is killed before it can handle the revert on code page, it goofs up the session.

> It might make sense to discuss this further in a separate issue.
I would be fine with that.

Reply:
As far as proposals go, this might be related, https://github.com/ziglang/zig/issues/3823 . If there were a buffer of some sort living in between for these things, then UTF16 decoding could live with using those. Also instead of returning a `File` it is proposing to return `Reader`s instead, which could work in the favor of having Windows return specific Console Reader/Writers for its stdio.

Reply:
Would it be possible to side-step the reader buffering problem by under-filling the read buffer, if the remaining space might not be large enough for a complete codepoint?

If the read ended in a leading surrogate, just make another `ReadFileW` or `ReadConsoleW` call to get the remaining UTF-16 word and complete the codepoint.

Reply:
Thanks for the detailed thoughts. Makes good sense. We definitely need a buffer to support `readByte`

One more naive question for you: Is there a reason we can't use a static (`threadlocal`?) variable for that buffer? It seems like for correct-ness, we'd actually prefer it to be global.

Reply:
> One more naive question for you: Is there a reason we can't use a static (`threadlocal`?) variable for that buffer? It seems like for correct-ness, we'd actually prefer it to be global.

I think that starts to dip into what the above [proposal](https://github.com/ziglang/zig/issues/3823) tries to potential provide as an answer from Andrew's comment. I'd be willing to throw in a `threadlocal` buffer+surrogate set for Windows reads and see what the looks like. Might come out cleaner for everyone. Give me a bit to toy with that (I'm not familiar with that keyword/it's usage 😅 ).

Reply:
Sounds great! Feel free to reach out on #stdlib-devel on the Discord if you'd talk through any of the finer points of the design 🙂

(I'm no concurrency expert either, but there are some folks in that channel who are)

Reply:
Gave it a shot, and at a glance, yea, it is operating as you thought 🦸 . It allows us to keep all changes within `File` without having to change any user facing APIs. `threadlocal` might be a decent call to have for any given stdin handle devs use, but I would think this might cause headaches if multiple threads are reading from the same handle on Windows, which I don't know of a good reason off the top of my head to do that anyways.

Let me update this PR to reflect the consolidation of all this + cleanup + comments.

Reply:
Alright, take a gander at this new version of `File` only changes to `read`/`write` and let me know what y'all think!

Reply:
Looks good to me overall.